go/worker/common/p2p/dispatch.go

```diff
@@ -0,0 +1,226 @@
+	// dispatch was successful?  Doing so ensures messages get
+	// propagated even if the local view is behind, but not doing so
+	// may be safer.
+
+	return true
+}
+
+func (h *topicHandler) dispatchMessage(peerID core.PeerID, m *queuedMsg, isInitial bool) error {
+	var err error
+	defer func() {
+		if err == nil || !isInitial {
+			return
+		}
+		if p2pError.IsPermanent(err) {
+			h.logger.Error("failed to dispatch message in-line, not retrying",
+				"err", err,
+				"peer_id", peerID,
+			)
+			return
+		}
+
+		// Kick off the retry worker if this is the initial attempt to
+		// dispatch the message.
+		for {
+			numWorkers := atomic.LoadUint64(&h.numWorkers)
+			if numWorkers > redispatchMaxWorkers {
+				h.logger.Error("failed to enqueue message for retry, queue full",
+					"peer_id", peerID,
+				)
+				return
+			}
+			if atomic.CompareAndSwapUint64(&h.numWorkers, numWorkers, numWorkers+1) {
+				h.logger.Error("failed to dispatch message in-line, retrying",
+					"err", err,
+					"peer_id", peerID,
+				)
+
+				go h.retryWorker(m)
+
+				return
+			}
+		}
+	}()
+
+	// Perhaps this should reject the message, but it is possible that
+	// the local node is just behind.  This does result in stale messages
+	// getting retried though.
+	if err = h.handler.AuthenticatePeer(m.from); err != nil {
+		return err
+	}
+
+	// Dispatch the message to the handler.
+	if err = h.handler.HandlePeerMessage(m.from, m.msg); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (h *topicHandler) retryWorker(m *queuedMsg) {
+	defer func() {
+		atomic.AddUint64(&h.numWorkers, ^uint64(0))
+	}()
+
+	off := backoff.WithMaxRetries(backoff.NewExponentialBackOff(), redispatchMaxRetries)
+	bctx := backoff.WithContext(off, h.ctx)
+
+	err := backoff.Retry(func() error {
+		derr := h.dispatchMessage(m.peerID, m, false)
+		switch derr {
+		case nil:
+			h.logger.Debug("succeeded in re-dispatching message",
+				"peer_id", m.peerID,
+			)
+		default:
+			if !p2pError.IsPermanent(derr) {
+				h.logger.Warn("failed to-redispatch message, will retry",
+					"err", derr,
+					"peer_id", m.peerID,
+				)
+			}
+		}
+		return derr
+	}, bctx)
+
+	if err != nil {
+		h.logger.Error("failed to re-dispatch message, not retrying",
+			"err", err,
+			"peer_id", m.peerID,
+		)
+	}
+}
+
+func newTopicHandler(p *P2P, runtimeID common.Namespace, handler Handler) (string, *topicHandler, error) {
+	topicID := runtimeIDToTopicID(runtimeID)
+	topic, err := p.pubsub.Join(topicID) // Note: Disallows duplicates.
+	if err != nil {
+		return "", nil, fmt.Errorf("worker/common/p2p: failed to join topic '%s': %w", topicID, err)
+	}
+
+	h := &topicHandler{
+		ctx:     p.ctx, // TODO: Should this support individual cancelation?
+		p2p:     p,
+		topic:   topic,
+		handler: handler,
+		logger:  logging.GetLogger("worker/common/p2p/" + topicID),
+	}
+	if h.cancelRelay, err = h.topic.Relay(); err != nil {
+		// This is non-fatal for now...
```

Comment:
As far as I can tell the only cases where this can happen is if the topic isn't currently actually subscribed, or if the pubsub context has been canceled.

Reply:
Maybe make it fatal in this case?

Reply:
Ok.